include/boost/rational.hpp

```diff
@@ -148,33 +152,102 @@ class rational :
 
     BOOST_CONSTEXPR
     rational() : num(0), den(1) {}
-    BOOST_CONSTEXPR
-    rational(param_type n) : num(n), den(1) {}
-    rational(param_type n, param_type d) : num(n), den(d) { normalize(); }
+    template <class T>
+    BOOST_CONSTEXPR rational(const T& n, typename enable_if_c<
+       rational_detail::is_compatible_integer<T, IntType>::value
+    >::type const* = 0) : num(n), den(1) {}
+    template <class T>
+    rational(const T& n, const T& d, typename enable_if_c<
+       rational_detail::is_compatible_integer<T, IntType>::value
+    >::type const* = 0) : num(n), den(d) {
+       normalize();
+    }
 
 #ifndef BOOST_NO_MEMBER_TEMPLATES
     template < typename NewType >
     BOOST_CONSTEXPR explicit
-    rational( rational<NewType> const &r )
-        : num( r.numerator() ), den( is_normalized(int_type( r.numerator() ),
-          int_type( r.denominator() )) ? r.denominator() :
-          throw bad_rational("bad rational: denormalized conversion") )
-    {}
+       rational(rational<NewType> const &r)
+       : num(r.numerator()), den(is_normalized(int_type(r.numerator()),
+       int_type(r.denominator())) ? r.denominator() :
+       (BOOST_THROW_EXCEPTION(bad_rational("bad rational: denormalized conversion")), 0)){}
 #endif
 
     // Default copy constructor and assignment are fine
 
     // Add assignment from IntType
-    rational& operator=(param_type i) { num = i; den = 1; return *this; }
+    template <class T>
+    typename enable_if_c<
+       rational_detail::is_compatible_integer<T, IntType>::value, rational &
+    >::type operator=(const T& n) { return assign(static_cast<IntType>(n), static_cast<IntType>(1)); }
 
     // Assign in place
-    rational& assign(param_type n, param_type d);
+    template <class T>
+    typename enable_if_c<
+       rational_detail::is_compatible_integer<T, IntType>::value, rational &
+    >::type assign(const T& n, const T& d)
+    {
+       return *this = rational<IntType>(static_cast<IntType>(n), static_cast<IntType>(d));
+    }
+    //
+    // The following overloads should probably *not* be provided - 
+    // but are provided for backwards compatibity reasons only.
+    // These allow for construction/assignment from types that
+    // are wider than IntType only if there is an implicit
+    // conversion from T to IntType, they will throw a bad_rational
+    // if the conversion results in loss of precision or undefined behaviour.
+    //
+    template <class T>
+    rational(const T& n, typename enable_if_c<
+       std::numeric_limits<T>::is_specialized && std::numeric_limits<T>::is_integer
+       && ((std::numeric_limits<T>::digits > std::numeric_limits<IntType>::digits)
+       || (std::numeric_limits<T>::is_signed != std::numeric_limits<IntType>::is_signed))
+       && (std::numeric_limits<T>::radix == std::numeric_limits<IntType>::radix)
+       && is_convertible<T, IntType>::value
+    >::type const* = 0)
```

Comment:
> This can be ambiguous with is_compatible_integer if T and IntType are 
> class types with an implicit narrowing conversion.
>
Nod.  I believe my changes for signedness conversions fix this also.